Fix get constraints with lvalue iterator input

efernandez opened the pull request:
This PR is on top of https://github.com/locusrobotics/fuse/pull/133, that extends the unit test for `getConstraints` to show a bug in the code that it's fixed here. See https://github.com/locusrobotics/fuse/pull/133 for more details.

efernandez committed:
Test getConstraints with lvalue iterator

This fails to change the `iterator` from the caller frame, since
all changes happen in the local copy of the iterator into the
`getConstraints` method.

efernandez committed:
Fix getConstraints for lvalue iterator

By taking a reference to the lvalue iterator input, the changes in
`getConstraints` can be seen from outside. Note that we also need to
assign the output of `std::copy` to `result`.

With the move semantics we make `getConstraints` work for both rvalue
and lvalue inputs, e.g. `std::back_inserter<>()` and iterators.

efernandez commented on fuse_constraints/include/fuse_constraints/variable_constraints.h:
```diff
@@ -98,7 +98,7 @@ class VariableConstraints
    */
   template <typename OutputIterator>
-  void getConstraints(const unsigned int variable_id, OutputIterator result) const;
+  void getConstraints(const unsigned int variable_id, OutputIterator&& result) const;
```
Maybe it makes more sense to have two implementations:

- One with `OutputIterator` that takes an rvalue by copy or with `OutputIterator&&` using move semantics. This will support `std::back_inserter`
-  Another one with `OutputIterator&` that takes an lvalue by reference. This will support https://github.com/locusrobotics/fuse/pull/133

In the case of `OutputIterator&` we also need `result = std::copy(...);`, but not in the other case.

Another option would be to change the `getConstraints` API to follow the one of `std::copy`, so it returns `result`, i.e. `return std::copy(...);`, in which case the `OutputIterator` arguments would always be passed by copy. :thinking:

svwilliams replied:
Based on the signature of `std::copy`, I think the correct thing to do is return the "one passed the last" iterator from the function.
```
template< class InputIt, class OutputIt >
OutputIt copy( InputIt first, InputIt last, OutputIt d_first );
```
And then fix up the usage in the marginalize code.

Thanks for catching this.

efernandez replied:
Done (in my last commit)

efernandez committed:
Improve unit test failure output

efernandez committed:
Re-order expected elimination order

efernandez commented on fuse_constraints/test/test_marginalize_variables.cpp:
```diff
@@ -175,14 +175,15 @@ TEST(MarginalizeVariables, ComputeEliminationOrder)
   auto expected = fuse_constraints::UuidOrdering();
   expected.push_back(x1->uuid());
   expected.push_back(x2->uuid());
-  expected.push_back(x3->uuid());
```
I had to change the expected order in this unit test. I believe the unit test was wrong, but I'm not sure how to find out the correct expected elimination order we should get here. Thoughts?

svwilliams replied:
I did a quick test using GTSAM's ordering system. The generated order is indeed: x1, x2, l1, x3

efernandez replied:
Cool. I'm just curious. How did you run the test in GTSAM? Is there any example code, or you had to build yours from scratch?

svwilliams replied:
I hacked up this unit test to reproduce our example.
https://github.com/borglab/gtsam/blob/develop/gtsam/inference/tests/testOrdering.cpp#L94

A "symbolic" factor graph models the structure of the graph without worrying about the numeric details of the cost functions. And GTSAM uses sequential numbers as variable names (at least at the symbolic level). So x1->0, x2->1, x3->2, l1->3, l2->4.
```
  SymbolicFactorGraph symbolicGraph;
  symbolicGraph.push_factor(0);
  symbolicGraph.push_factor(0, 1);
  symbolicGraph.push_factor(1, 2);
  symbolicGraph.push_factor(0, 3);
  symbolicGraph.push_factor(1, 3);
  symbolicGraph.push_factor(2, 4);
```
The above reproduces the graph structure in our unittest. 

And then we can generate an ordering using COLAMD:
```
  Ordering actual = Ordering::ColamdConstrainedFirst(symbolicGraph, list_of(1)(0));
```
The second parameter is the list of variables to constrain to the front of the ordering.

And the actual ordering passes this test:
```
  Ordering expected = list_of(0)(1)(3)(2)(4);
  EXPECT(assert_equal(expected, actual));
```
which is (x1, x2, l1, x3, l2).

Note that we are actually computing an ordering for just some of the graph...anything that directly connects to x1 or x2. That includes everything except the x3--l2 factor. If we comment out `symbolicGraph.push_factor(2, 4);`, then we reproduce what we are actually computing. And that returns (0, 1, 3, 2) or (x1, x2, l1, x3).

So both situations produce the (x1, x2, l1, x3) ordering, which makes it pretty easy to decide that is correct. 

I contributed to GTSAM for a couple of years while attending GaTech, which is why I know how all this works.

efernandez replied:
Thanks. Awesome!

That looks good and makes sense.

efernandez commented on fuse_constraints/test/test_marginalize_variables.cpp:
```diff
@@ -175,14 +175,15 @@ TEST(MarginalizeVariables, ComputeEliminationOrder)
   for (size_t i = 0; i < expected.size(); ++i)
   {
-    EXPECT_EQ(expected.at(i), actual.at(i));
+    SCOPED_TRACE(i);
```
This simply makes the output easier to read and find out which iteration failed, so instead of
``` bash
/home/efernandez/dev/ws/cpr_ws/src/fuse/fuse_constraints/test/test_marginalize_variables.cpp:186: Failure
Value of: actual.at(i)
  Actual: { '\x93' (147), '\xD1' (209), '\xE4' (228), 'u' (117, 0x75), '\xCC' (204), '?' (63, 0x3F), 'H' (72, 0x48), '\xEC' (236), '\xBD' (189), '\x18' (24), '0' (48, 0x30), 'G' (71, 0x47), '\xBE' (190), '\x9E' (158), '\xA5' (165), '\x81' (129) }
Expected: expected.at(i)
Which is: { '\xE5' (229), '\x8B' (139), 'r' (114, 0x72), 'b' (98, 0x62), 'V' (86, 0x56), '\xE5' (229), 'C' (67, 0x43), '\xE9' (233), '\x8E' (142), '|' (124, 0x7C), '\x9D' (157), '\xA3' (163), '7' (55, 0x37), '\xC8' (200), 'O' (79, 0x4F), '\xD5' (213) }
/home/efernandez/dev/ws/cpr_ws/src/fuse/fuse_constraints/test/test_marginalize_variables.cpp:186: Failure
Value of: actual.at(i)
  Actual: { '\xE5' (229), '\x8B' (139), 'r' (114, 0x72), 'b' (98, 0x62), 'V' (86, 0x56), '\xE5' (229), 'C' (67, 0x43), '\xE9' (233), '\x8E' (142), '|' (124, 0x7C), '\x9D' (157), '\xA3' (163), '7' (55, 0x37), '\xC8' (200), 'O' (79, 0x4F), '\xD5' (213) }
Expected: expected.at(i)
Which is: { '\x93' (147), '\xD1' (209), '\xE4' (228), 'u' (117, 0x75), '\xCC' (204), '?' (63, 0x3F), 'H' (72, 0x48), '\xEC' (236), '\xBD' (189), '\x18' (24), '0' (48, 0x30), 'G' (71, 0x47), '\xBE' (190), '\x9E' (158), '\xA5' (165), '\x81' (129) }
```
we get this:
``` bash
/home/efernandez/dev/ws/cpr_ws/src/fuse/fuse_constraints/test/test_marginalize_variables.cpp:186: Failure
Value of: fuse_core::uuid::to_string(actual.at(i))
  Actual: "29e83c8c-24cc-4e1a-bbde-64ffa628a9bb"
Expected: fuse_core::uuid::to_string(expected.at(i))
Which is: "bcfcb97c-59c5-4291-9caa-11ad7141ac23"
Google Test trace:
/home/efernandez/dev/ws/cpr_ws/src/fuse/fuse_constraints/test/test_marginalize_variables.cpp:185: 2
/home/efernandez/dev/ws/cpr_ws/src/fuse/fuse_constraints/test/test_marginalize_variables.cpp:186: Failure
Value of: fuse_core::uuid::to_string(actual.at(i))
  Actual: "bcfcb97c-59c5-4291-9caa-11ad7141ac23"
Expected: fuse_core::uuid::to_string(expected.at(i))
Which is: "29e83c8c-24cc-4e1a-bbde-64ffa628a9bb"
Google Test trace:
/home/efernandez/dev/ws/cpr_ws/src/fuse/fuse_constraints/test/test_marginalize_variables.cpp:185: 3
```

svwilliams replied:
What are you talking about? It was perfectly easy to read before. :)

efernandez committed:
Change getConstraints signature to return iterator

efernandez commented:
@svwilliams Could you paste here the error output from the PR checks, please? I believe with my last commits they should have passed, including `roslint`.

```bash
~/dev/ws/cpr_ws/build/fuse_constraints$ make run_tests
[  4%] Built target gtest
[  4%] Built target geometry_msgs_generate_messages_lisp
[  4%] Built target std_msgs_generate_messages_lisp
[  4%] Built target roscpp_generate_messages_py
[  4%] Built target geometry_msgs_generate_messages_py
[  4%] Built target fuse_msgs_generate_messages_cpp
[  4%] Built target rosgraph_msgs_generate_messages_lisp
[  4%] Built target rosgraph_msgs_generate_messages_py
[  4%] Built target fuse_msgs_generate_messages_py
[  4%] Built target rosgraph_msgs_generate_messages_cpp
[  4%] Built target std_msgs_generate_messages_cpp
[  4%] Built target std_msgs_generate_messages_py
[  4%] Built target fuse_msgs_generate_messages_lisp
[  4%] Built target roscpp_generate_messages_cpp
[  4%] Built target roscpp_generate_messages_lisp
[  4%] Built target geometry_msgs_generate_messages_cpp
Scanning dependencies of target fuse_constraints
[  6%] Building CXX object CMakeFiles/fuse_constraints.dir/src/absolute_constraint.cpp.o
[  9%] Building CXX object CMakeFiles/fuse_constraints.dir/src/absolute_orientation_3d_stamped_constraint.cpp.o
[ 11%] Building CXX object CMakeFiles/fuse_constraints.dir/src/absolute_orientation_3d_stamped_euler_constraint.cpp.o
[ 13%] Building CXX object CMakeFiles/fuse_constraints.dir/src/absolute_pose_2d_stamped_constraint.cpp.o
[ 15%] Building CXX object CMakeFiles/fuse_constraints.dir/src/absolute_pose_3d_stamped_constraint.cpp.o
[ 18%] Building CXX object CMakeFiles/fuse_constraints.dir/src/marginal_constraint.cpp.o
[ 20%] Building CXX object CMakeFiles/fuse_constraints.dir/src/marginal_cost_function.cpp.o
[ 22%] Building CXX object CMakeFiles/fuse_constraints.dir/src/marginalize_variables.cpp.o
[ 25%] Building CXX object CMakeFiles/fuse_constraints.dir/src/normal_delta.cpp.o
[ 27%] Building CXX object CMakeFiles/fuse_constraints.dir/src/normal_delta_orientation_2d.cpp.o
[ 29%] Building CXX object CMakeFiles/fuse_constraints.dir/src/normal_prior_orientation_2d.cpp.o
[ 31%] Building CXX object CMakeFiles/fuse_constraints.dir/src/relative_constraint.cpp.o
[ 34%] Building CXX object CMakeFiles/fuse_constraints.dir/src/relative_orientation_3d_stamped_constraint.cpp.o
[ 36%] Building CXX object CMakeFiles/fuse_constraints.dir/src/relative_pose_2d_stamped_constraint.cpp.o
[ 38%] Building CXX object CMakeFiles/fuse_constraints.dir/src/relative_pose_3d_stamped_constraint.cpp.o
[ 40%] Building CXX object CMakeFiles/fuse_constraints.dir/src/uuid_ordering.cpp.o
[ 43%] Building CXX object CMakeFiles/fuse_constraints.dir/src/variable_constraints.cpp.o
[ 45%] Linking CXX shared library /home/efernandez/dev/ws/cpr_ws/devel/.private/fuse_constraints/lib/libfuse_constraints.so
[ 45%] Built target fuse_constraints
Scanning dependencies of target test_absolute_pose_3d_stamped_constraint
[ 47%] Building CXX object CMakeFiles/test_absolute_pose_3d_stamped_constraint.dir/test/test_absolute_pose_3d_stamped_constraint.cpp.o
[ 50%] Linking CXX executable /home/efernandez/dev/ws/cpr_ws/devel/.private/fuse_constraints/lib/fuse_constraints/test_absolute_pose_3d_stamped_constraint
[ 50%] Built target test_absolute_pose_3d_stamped_constraint
Scanning dependencies of target test_relative_pose_3d_stamped_constraint
[ 52%] Building CXX object CMakeFiles/test_relative_pose_3d_stamped_constraint.dir/test/test_relative_pose_3d_stamped_constraint.cpp.o
[ 54%] Linking CXX executable /home/efernandez/dev/ws/cpr_ws/devel/.private/fuse_constraints/lib/fuse_constraints/test_relative_pose_3d_stamped_constraint
[ 54%] Built target test_relative_pose_3d_stamped_constraint
Scanning dependencies of target test_relative_pose_2d_stamped_constraint
[ 56%] Building CXX object CMakeFiles/test_relative_pose_2d_stamped_constraint.dir/test/test_relative_pose_2d_stamped_constraint.cpp.o
[ 59%] Linking CXX executable /home/efernandez/dev/ws/cpr_ws/devel/.private/fuse_constraints/lib/fuse_constraints/test_relative_pose_2d_stamped_constraint
[ 59%] Built target test_relative_pose_2d_stamped_constraint
Scanning dependencies of target test_marginal_constraint
[ 61%] Building CXX object CMakeFiles/test_marginal_constraint.dir/test/test_marginal_constraint.cpp.o
[ 63%] Linking CXX executable /home/efernandez/dev/ws/cpr_ws/devel/.private/fuse_constraints/lib/fuse_constraints/test_marginal_constraint
[ 63%] Built target test_marginal_constraint
Scanning dependencies of target test_relative_constraint
[ 65%] Building CXX object CMakeFiles/test_relative_constraint.dir/test/test_relative_constraint.cpp.o
[ 68%] Linking CXX executable /home/efernandez/dev/ws/cpr_ws/devel/.private/fuse_constraints/lib/fuse_constraints/test_relative_constraint
[ 68%] Built target test_relative_constraint
Scanning dependencies of target test_marginalize_variables
[ 70%] Building CXX object CMakeFiles/test_marginalize_variables.dir/test/test_marginalize_variables.cpp.o
[ 72%] Linking CXX executable /home/efernandez/dev/ws/cpr_ws/devel/.private/fuse_constraints/lib/fuse_constraints/test_marginalize_variables
[ 72%] Built target test_marginalize_variables
Scanning dependencies of target test_absolute_orientation_3d_stamped_constraint
[ 75%] Building CXX object CMakeFiles/test_absolute_orientation_3d_stamped_constraint.dir/test/test_absolute_orientation_3d_stamped_constraint.cpp.o
[ 77%] Linking CXX executable /home/efernandez/dev/ws/cpr_ws/devel/.private/fuse_constraints/lib/fuse_constraints/test_absolute_orientation_3d_stamped_constraint
[ 77%] Built target test_absolute_orientation_3d_stamped_constraint
Scanning dependencies of target test_absolute_constraint
[ 79%] Building CXX object CMakeFiles/test_absolute_constraint.dir/test/test_absolute_constraint.cpp.o
[ 81%] Linking CXX executable /home/efernandez/dev/ws/cpr_ws/devel/.private/fuse_constraints/lib/fuse_constraints/test_absolute_constraint
[ 81%] Built target test_absolute_constraint
Scanning dependencies of target test_absolute_orientation_3d_stamped_euler_constraint
[ 84%] Building CXX object CMakeFiles/test_absolute_orientation_3d_stamped_euler_constraint.dir/test/test_absolute_orientation_3d_stamped_euler_constraint.cpp.o
[ 86%] Linking CXX executable /home/efernandez/dev/ws/cpr_ws/devel/.private/fuse_constraints/lib/fuse_constraints/test_absolute_orientation_3d_stamped_euler_constraint
[ 86%] Built target test_absolute_orientation_3d_stamped_euler_constraint
Scanning dependencies of target test_absolute_pose_2d_stamped_constraint
[ 88%] Building CXX object CMakeFiles/test_absolute_pose_2d_stamped_constraint.dir/test/test_absolute_pose_2d_stamped_constraint.cpp.o
[ 90%] Linking CXX executable /home/efernandez/dev/ws/cpr_ws/devel/.private/fuse_constraints/lib/fuse_constraints/test_absolute_pose_2d_stamped_constraint
[ 90%] Built target test_absolute_pose_2d_stamped_constraint
Scanning dependencies of target test_variable_constraints
[ 93%] Building CXX object CMakeFiles/test_variable_constraints.dir/test/test_variable_constraints.cpp.o
[ 95%] Linking CXX executable /home/efernandez/dev/ws/cpr_ws/devel/.private/fuse_constraints/lib/fuse_constraints/test_variable_constraints
[ 95%] Built target test_variable_constraints
Scanning dependencies of target test_uuid_ordering
[ 97%] Building CXX object CMakeFiles/test_uuid_ordering.dir/test/test_uuid_ordering.cpp.o
[100%] Linking CXX executable /home/efernandez/dev/ws/cpr_ws/devel/.private/fuse_constraints/lib/fuse_constraints/test_uuid_ordering
[100%] Built target test_uuid_ordering
[100%] Built target tests
Removing test result files from '/home/efernandez/dev/ws/cpr_ws/build/fuse_constraints/test_results/fuse_constraints'
- removing '/home/efernandez/dev/ws/cpr_ws/build/fuse_constraints/test_results/fuse_constraints/gtest-test_relative_pose_3d_stamped_constraint.xml'
- removing '/home/efernandez/dev/ws/cpr_ws/build/fuse_constraints/test_results/fuse_constraints/gtest-test_absolute_pose_3d_stamped_constraint.xml'
- removing '/home/efernandez/dev/ws/cpr_ws/build/fuse_constraints/test_results/fuse_constraints/gtest-test_relative_pose_2d_stamped_constraint.xml'
- removing '/home/efernandez/dev/ws/cpr_ws/build/fuse_constraints/test_results/fuse_constraints/gtest-test_absolute_constraint.xml'
- removing '/home/efernandez/dev/ws/cpr_ws/build/fuse_constraints/test_results/fuse_constraints/gtest-test_marginal_constraint.xml'
- removing '/home/efernandez/dev/ws/cpr_ws/build/fuse_constraints/test_results/fuse_constraints/gtest-test_absolute_orientation_3d_stamped_euler_constraint.xml'
- removing '/home/efernandez/dev/ws/cpr_ws/build/fuse_constraints/test_results/fuse_constraints/gtest-test_absolute_orientation_3d_stamped_constraint.xml'
- removing '/home/efernandez/dev/ws/cpr_ws/build/fuse_constraints/test_results/fuse_constraints/roslint-fuse_constraints.xml'
- removing '/home/efernandez/dev/ws/cpr_ws/build/fuse_constraints/test_results/fuse_constraints/gtest-test_variable_constraints.xml'
- removing '/home/efernandez/dev/ws/cpr_ws/build/fuse_constraints/test_results/fuse_constraints/gtest-test_marginalize_variables.xml'
- removing '/home/efernandez/dev/ws/cpr_ws/build/fuse_constraints/test_results/fuse_constraints/gtest-test_relative_constraint.xml'
- removing '/home/efernandez/dev/ws/cpr_ws/build/fuse_constraints/test_results/fuse_constraints/gtest-test_absolute_pose_2d_stamped_constraint.xml'
- removing '/home/efernandez/dev/ws/cpr_ws/build/fuse_constraints/test_results/fuse_constraints/gtest-test_uuid_ordering.xml'
[100%] Built target clean_test_results_fuse_constraints
-- run_tests.py: execute commands
  /home/efernandez/dev/ws/cpr_ws/devel/.private/fuse_constraints/lib/fuse_constraints/test_variable_constraints --gtest_output=xml:/home/efernandez/dev/ws/cpr_ws/build/fuse_constraints/test_results/fuse_constraints/gtest-test_variable_constraints.xml
[==========] Running 3 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 3 tests from VariableConstraints
[ RUN      ] VariableConstraints.Size
[       OK ] VariableConstraints.Size (0 ms)
[ RUN      ] VariableConstraints.NextVariableIndex
[       OK ] VariableConstraints.NextVariableIndex (0 ms)
[ RUN      ] VariableConstraints.GetConstraints
[       OK ] VariableConstraints.GetConstraints (0 ms)
[----------] 3 tests from VariableConstraints (0 ms total)

[----------] Global test environment tear-down
[==========] 3 tests from 1 test case ran. (1 ms total)
[  PASSED  ] 3 tests.
-- run_tests.py: verify result "/home/efernandez/dev/ws/cpr_ws/build/fuse_constraints/test_results/fuse_constraints/gtest-test_variable_constraints.xml"
[100%] Built target _run_tests_fuse_constraints_gtest_test_variable_constraints
-- run_tests.py: execute commands
  /home/efernandez/dev/ws/cpr_ws/devel/.private/fuse_constraints/lib/fuse_constraints/test_relative_pose_3d_stamped_constraint --gtest_output=xml:/home/efernandez/dev/ws/cpr_ws/build/fuse_constraints/test_results/fuse_constraints/gtest-test_relative_pose_3d_stamped_constraint.xml
[==========] Running 4 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 4 tests from RelativePose3DStampedConstraint
[ RUN      ] RelativePose3DStampedConstraint.Constructor
[       OK ] RelativePose3DStampedConstraint.Constructor (0 ms)
[ RUN      ] RelativePose3DStampedConstraint.Covariance
[       OK ] RelativePose3DStampedConstraint.Covariance (0 ms)
[ RUN      ] RelativePose3DStampedConstraint.Optimization
[       OK ] RelativePose3DStampedConstraint.Optimization (4 ms)
[ RUN      ] RelativePose3DStampedConstraint.Serialization
[       OK ] RelativePose3DStampedConstraint.Serialization (0 ms)
[----------] 4 tests from RelativePose3DStampedConstraint (5 ms total)

[----------] Global test environment tear-down
[==========] 4 tests from 1 test case ran. (5 ms total)
[  PASSED  ] 4 tests.
-- run_tests.py: verify result "/home/efernandez/dev/ws/cpr_ws/build/fuse_constraints/test_results/fuse_constraints/gtest-test_relative_pose_3d_stamped_constraint.xml"
[100%] Built target _run_tests_fuse_constraints_gtest_test_relative_pose_3d_stamped_constraint
-- run_tests.py: execute commands
  /home/efernandez/dev/ws/cpr_ws/devel/.private/fuse_constraints/lib/fuse_constraints/test_marginalize_variables --gtest_output=xml:/home/efernandez/dev/ws/cpr_ws/build/fuse_constraints/test_results/fuse_constraints/gtest-test_marginalize_variables.xml
[==========] Running 4 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 4 tests from MarginalizeVariables
[ RUN      ] MarginalizeVariables.ComputeEliminationOrder
[       OK ] MarginalizeVariables.ComputeEliminationOrder (0 ms)
[ RUN      ] MarginalizeVariables.Linearize
[       OK ] MarginalizeVariables.Linearize (0 ms)
[ RUN      ] MarginalizeVariables.MarginalizeNext
[       OK ] MarginalizeVariables.MarginalizeNext (1 ms)
[ RUN      ] MarginalizeVariables.MarginalizeVariables
[       OK ] MarginalizeVariables.MarginalizeVariables (3 ms)
[----------] 4 tests from MarginalizeVariables (4 ms total)

[----------] Global test environment tear-down
[==========] 4 tests from 1 test case ran. (5 ms total)
[  PASSED  ] 4 tests.
-- run_tests.py: verify result "/home/efernandez/dev/ws/cpr_ws/build/fuse_constraints/test_results/fuse_constraints/gtest-test_marginalize_variables.xml"
[100%] Built target _run_tests_fuse_constraints_gtest_test_marginalize_variables
-- run_tests.py: execute commands
  /home/efernandez/dev/ws/cpr_ws/devel/.private/fuse_constraints/lib/fuse_constraints/test_marginal_constraint --gtest_output=xml:/home/efernandez/dev/ws/cpr_ws/build/fuse_constraints/test_results/fuse_constraints/gtest-test_marginal_constraint.xml
[==========] Running 4 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 4 tests from MarginalConstraint
[ RUN      ] MarginalConstraint.OneVariable
[       OK ] MarginalConstraint.OneVariable (0 ms)
[ RUN      ] MarginalConstraint.TwoVariables
[       OK ] MarginalConstraint.TwoVariables (0 ms)
[ RUN      ] MarginalConstraint.LocalParameterization
[       OK ] MarginalConstraint.LocalParameterization (0 ms)
[ RUN      ] MarginalConstraint.Serialization
[       OK ] MarginalConstraint.Serialization (0 ms)
[----------] 4 tests from MarginalConstraint (0 ms total)

[----------] Global test environment tear-down
[==========] 4 tests from 1 test case ran. (1 ms total)
[  PASSED  ] 4 tests.
-- run_tests.py: verify result "/home/efernandez/dev/ws/cpr_ws/build/fuse_constraints/test_results/fuse_constraints/gtest-test_marginal_constraint.xml"
[100%] Built target _run_tests_fuse_constraints_gtest_test_marginal_constraint
-- run_tests.py: execute commands
  /home/efernandez/dev/ws/cpr_ws/devel/.private/fuse_constraints/lib/fuse_constraints/test_relative_constraint --gtest_output=xml:/home/efernandez/dev/ws/cpr_ws/build/fuse_constraints/test_results/fuse_constraints/gtest-test_relative_constraint.xml
[==========] Running 6 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 6 tests from RelativeConstraint
[ RUN      ] RelativeConstraint.Constructor
[       OK ] RelativeConstraint.Constructor (0 ms)
[ RUN      ] RelativeConstraint.PartialMeasurement
[       OK ] RelativeConstraint.PartialMeasurement (0 ms)
[ RUN      ] RelativeConstraint.Covariance
[       OK ] RelativeConstraint.Covariance (0 ms)
[ RUN      ] RelativeConstraint.Optimization
[       OK ] RelativeConstraint.Optimization (18 ms)
[ RUN      ] RelativeConstraint.RelativeOrientation2DOptimization
[       OK ] RelativeConstraint.RelativeOrientation2DOptimization (3 ms)
[ RUN      ] RelativeConstraint.Serialization
[       OK ] RelativeConstraint.Serialization (0 ms)
[----------] 6 tests from RelativeConstraint (21 ms total)

[----------] Global test environment tear-down
[==========] 6 tests from 1 test case ran. (23 ms total)
[  PASSED  ] 6 tests.
-- run_tests.py: verify result "/home/efernandez/dev/ws/cpr_ws/build/fuse_constraints/test_results/fuse_constraints/gtest-test_relative_constraint.xml"
[100%] Built target _run_tests_fuse_constraints_gtest_test_relative_constraint
-- run_tests.py: execute commands
  /home/efernandez/dev/ws/cpr_ws/devel/.private/fuse_constraints/lib/fuse_constraints/test_absolute_pose_3d_stamped_constraint --gtest_output=xml:/home/efernandez/dev/ws/cpr_ws/build/fuse_constraints/test_results/fuse_constraints/gtest-test_absolute_pose_3d_stamped_constraint.xml
[==========] Running 4 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 4 tests from AbsolutePose3DStampedConstraint
[ RUN      ] AbsolutePose3DStampedConstraint.Constructor
[       OK ] AbsolutePose3DStampedConstraint.Constructor (0 ms)
[ RUN      ] AbsolutePose3DStampedConstraint.Covariance
[       OK ] AbsolutePose3DStampedConstraint.Covariance (0 ms)
[ RUN      ] AbsolutePose3DStampedConstraint.Optimization
[       OK ] AbsolutePose3DStampedConstraint.Optimization (6 ms)
[ RUN      ] AbsolutePose3DStampedConstraint.Serialization
[       OK ] AbsolutePose3DStampedConstraint.Serialization (0 ms)
[----------] 4 tests from AbsolutePose3DStampedConstraint (6 ms total)

[----------] Global test environment tear-down
[==========] 4 tests from 1 test case ran. (6 ms total)
[  PASSED  ] 4 tests.
-- run_tests.py: verify result "/home/efernandez/dev/ws/cpr_ws/build/fuse_constraints/test_results/fuse_constraints/gtest-test_absolute_pose_3d_stamped_constraint.xml"
[100%] Built target _run_tests_fuse_constraints_gtest_test_absolute_pose_3d_stamped_constraint
-- run_tests.py: execute commands
  /home/efernandez/dev/ws/cpr_ws/devel/.private/fuse_constraints/lib/fuse_constraints/test_uuid_ordering --gtest_output=xml:/home/efernandez/dev/ws/cpr_ws/build/fuse_constraints/test_results/fuse_constraints/gtest-test_uuid_ordering.xml
[==========] Running 5 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 5 tests from UuidOrdering
[ RUN      ] UuidOrdering.Constructor
[       OK ] UuidOrdering.Constructor (0 ms)
[ RUN      ] UuidOrdering.Access
[       OK ] UuidOrdering.Access (0 ms)
[ RUN      ] UuidOrdering.PushBack
[       OK ] UuidOrdering.PushBack (0 ms)
[ RUN      ] UuidOrdering.Size
[       OK ] UuidOrdering.Size (0 ms)
[ RUN      ] UuidOrdering.Exists
[       OK ] UuidOrdering.Exists (0 ms)
[----------] 5 tests from UuidOrdering (0 ms total)

[----------] Global test environment tear-down
[==========] 5 tests from 1 test case ran. (1 ms total)
[  PASSED  ] 5 tests.
-- run_tests.py: verify result "/home/efernandez/dev/ws/cpr_ws/build/fuse_constraints/test_results/fuse_constraints/gtest-test_uuid_ordering.xml"
[100%] Built target _run_tests_fuse_constraints_gtest_test_uuid_ordering
-- run_tests.py: execute commands
  /home/efernandez/dev/ws/cpr_ws/devel/.private/fuse_constraints/lib/fuse_constraints/test_relative_pose_2d_stamped_constraint --gtest_output=xml:/home/efernandez/dev/ws/cpr_ws/build/fuse_constraints/test_results/fuse_constraints/gtest-test_relative_pose_2d_stamped_constraint.xml
[==========] Running 5 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 5 tests from RelativePose2DStampedConstraint
[ RUN      ] RelativePose2DStampedConstraint.Constructor
[       OK ] RelativePose2DStampedConstraint.Constructor (0 ms)
[ RUN      ] RelativePose2DStampedConstraint.Covariance
[       OK ] RelativePose2DStampedConstraint.Covariance (0 ms)
[ RUN      ] RelativePose2DStampedConstraint.OptimizationFull
[       OK ] RelativePose2DStampedConstraint.OptimizationFull (9 ms)
[ RUN      ] RelativePose2DStampedConstraint.OptimizationPartial
[       OK ] RelativePose2DStampedConstraint.OptimizationPartial (9 ms)
[ RUN      ] RelativePose2DStampedConstraint.Serialization
[       OK ] RelativePose2DStampedConstraint.Serialization (0 ms)
[----------] 5 tests from RelativePose2DStampedConstraint (18 ms total)

[----------] Global test environment tear-down
[==========] 5 tests from 1 test case ran. (18 ms total)
[  PASSED  ] 5 tests.
-- run_tests.py: verify result "/home/efernandez/dev/ws/cpr_ws/build/fuse_constraints/test_results/fuse_constraints/gtest-test_relative_pose_2d_stamped_constraint.xml"
[100%] Built target _run_tests_fuse_constraints_gtest_test_relative_pose_2d_stamped_constraint
-- run_tests.py: execute commands
  /home/efernandez/dev/ws/cpr_ws/devel/.private/fuse_constraints/lib/fuse_constraints/test_absolute_orientation_3d_stamped_euler_constraint --gtest_output=xml:/home/efernandez/dev/ws/cpr_ws/build/fuse_constraints/test_results/fuse_constraints/gtest-test_absolute_orientation_3d_stamped_euler_constraint.xml
[==========] Running 5 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 5 tests from AbsoluteOrientation3DStampedEulerConstraint
[ RUN      ] AbsoluteOrientation3DStampedEulerConstraint.Constructor
[       OK ] AbsoluteOrientation3DStampedEulerConstraint.Constructor (0 ms)
[ RUN      ] AbsoluteOrientation3DStampedEulerConstraint.Covariance
[       OK ] AbsoluteOrientation3DStampedEulerConstraint.Covariance (0 ms)
[ RUN      ] AbsoluteOrientation3DStampedEulerConstraint.OptimizationFull
[       OK ] AbsoluteOrientation3DStampedEulerConstraint.OptimizationFull (1 ms)
[ RUN      ] AbsoluteOrientation3DStampedEulerConstraint.OptimizationPartial
[       OK ] AbsoluteOrientation3DStampedEulerConstraint.OptimizationPartial (1 ms)
[ RUN      ] AbsoluteOrientation3DStampedEulerConstraint.Serialization
[       OK ] AbsoluteOrientation3DStampedEulerConstraint.Serialization (0 ms)
[----------] 5 tests from AbsoluteOrientation3DStampedEulerConstraint (2 ms total)

[----------] Global test environment tear-down
[==========] 5 tests from 1 test case ran. (2 ms total)
[  PASSED  ] 5 tests.
-- run_tests.py: verify result "/home/efernandez/dev/ws/cpr_ws/build/fuse_constraints/test_results/fuse_constraints/gtest-test_absolute_orientation_3d_stamped_euler_constraint.xml"
[100%] Built target _run_tests_fuse_constraints_gtest_test_absolute_orientation_3d_stamped_euler_constraint
-- run_tests.py: execute commands
  /home/efernandez/dev/ws/cpr_ws/devel/.private/fuse_constraints/lib/fuse_constraints/test_absolute_orientation_3d_stamped_constraint --gtest_output=xml:/home/efernandez/dev/ws/cpr_ws/build/fuse_constraints/test_results/fuse_constraints/gtest-test_absolute_orientation_3d_stamped_constraint.xml
[==========] Running 4 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 4 tests from AbsoluteOrientation3DStampedConstraint
[ RUN      ] AbsoluteOrientation3DStampedConstraint.Constructor
[       OK ] AbsoluteOrientation3DStampedConstraint.Constructor (0 ms)
[ RUN      ] AbsoluteOrientation3DStampedConstraint.Covariance
[       OK ] AbsoluteOrientation3DStampedConstraint.Covariance (0 ms)
[ RUN      ] AbsoluteOrientation3DStampedConstraint.Optimization
[       OK ] AbsoluteOrientation3DStampedConstraint.Optimization (4 ms)
[ RUN      ] AbsoluteOrientation3DStampedConstraint.Serialization
[       OK ] AbsoluteOrientation3DStampedConstraint.Serialization (0 ms)
[----------] 4 tests from AbsoluteOrientation3DStampedConstraint (4 ms total)

[----------] Global test environment tear-down
[==========] 4 tests from 1 test case ran. (4 ms total)
[  PASSED  ] 4 tests.
-- run_tests.py: verify result "/home/efernandez/dev/ws/cpr_ws/build/fuse_constraints/test_results/fuse_constraints/gtest-test_absolute_orientation_3d_stamped_constraint.xml"
[100%] Built target _run_tests_fuse_constraints_gtest_test_absolute_orientation_3d_stamped_constraint
-- run_tests.py: execute commands
  /home/efernandez/dev/ws/cpr_ws/devel/.private/fuse_constraints/lib/fuse_constraints/test_absolute_constraint --gtest_output=xml:/home/efernandez/dev/ws/cpr_ws/build/fuse_constraints/test_results/fuse_constraints/gtest-test_absolute_constraint.xml
[==========] Running 7 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 7 tests from AbsoluteConstraint
[ RUN      ] AbsoluteConstraint.Constructor
[       OK ] AbsoluteConstraint.Constructor (0 ms)
[ RUN      ] AbsoluteConstraint.PartialMeasurement
[       OK ] AbsoluteConstraint.PartialMeasurement (0 ms)
[ RUN      ] AbsoluteConstraint.Covariance
[       OK ] AbsoluteConstraint.Covariance (0 ms)
[ RUN      ] AbsoluteConstraint.Optimization
[       OK ] AbsoluteConstraint.Optimization (9 ms)
[ RUN      ] AbsoluteConstraint.PartialOptimization
[       OK ] AbsoluteConstraint.PartialOptimization (0 ms)
[ RUN      ] AbsoluteConstraint.AbsoluteOrientation2DOptimization
[       OK ] AbsoluteConstraint.AbsoluteOrientation2DOptimization (1 ms)
[ RUN      ] AbsoluteConstraint.Serialization
[       OK ] AbsoluteConstraint.Serialization (0 ms)
[----------] 7 tests from AbsoluteConstraint (11 ms total)

[----------] Global test environment tear-down
[==========] 7 tests from 1 test case ran. (11 ms total)
[  PASSED  ] 7 tests.
-- run_tests.py: verify result "/home/efernandez/dev/ws/cpr_ws/build/fuse_constraints/test_results/fuse_constraints/gtest-test_absolute_constraint.xml"
[100%] Built target _run_tests_fuse_constraints_gtest_test_absolute_constraint
-- run_tests.py: execute commands
  /home/efernandez/dev/ws/cpr_ws/devel/.private/fuse_constraints/lib/fuse_constraints/test_absolute_pose_2d_stamped_constraint --gtest_output=xml:/home/efernandez/dev/ws/cpr_ws/build/fuse_constraints/test_results/fuse_constraints/gtest-test_absolute_pose_2d_stamped_constraint.xml
[==========] Running 5 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 5 tests from AbsolutePose2DStampedConstraint
[ RUN      ] AbsolutePose2DStampedConstraint.Constructor
[       OK ] AbsolutePose2DStampedConstraint.Constructor (0 ms)
[ RUN      ] AbsolutePose2DStampedConstraint.Covariance
[       OK ] AbsolutePose2DStampedConstraint.Covariance (1 ms)
[ RUN      ] AbsolutePose2DStampedConstraint.OptimizationFull
[       OK ] AbsolutePose2DStampedConstraint.OptimizationFull (9 ms)
[ RUN      ] AbsolutePose2DStampedConstraint.OptimizationPartial
[       OK ] AbsolutePose2DStampedConstraint.OptimizationPartial (8 ms)
[ RUN      ] AbsolutePose2DStampedConstraint.Serialization
[       OK ] AbsolutePose2DStampedConstraint.Serialization (0 ms)
[----------] 5 tests from AbsolutePose2DStampedConstraint (18 ms total)

[----------] Global test environment tear-down
[==========] 5 tests from 1 test case ran. (18 ms total)
[  PASSED  ] 5 tests.
-- run_tests.py: verify result "/home/efernandez/dev/ws/cpr_ws/build/fuse_constraints/test_results/fuse_constraints/gtest-test_absolute_pose_2d_stamped_constraint.xml"
[100%] Built target _run_tests_fuse_constraints_gtest_test_absolute_pose_2d_stamped_constraint
[100%] Built target _run_tests_fuse_constraints_gtest
-- run_tests.py: execute commands with working directory "/home/efernandez/dev/ws/cpr_ws/build/fuse_constraints"
  /opt/clearpath/2.18/share/roslint/cmake/../../../lib/roslint/test_wrapper /home/efernandez/dev/ws/cpr_ws/build/fuse_constraints/test_results/fuse_constraints/roslint-fuse_constraints.xml make roslint_fuse_constraints
-- run_tests.py: verify result "/home/efernandez/dev/ws/cpr_ws/build/fuse_constraints/test_results/fuse_constraints/roslint-fuse_constraints.xml"
Built target _run_tests_fuse_constraints_roslint_package
Built target _run_tests_fuse_constraints_roslint
Built target _run_tests_fuse_constraints
Built target run_tests
```

svwilliams commented:
```
[2020-01-22T16:33:28.347Z] /home/ec2-user/workspace/test_rosdistro_fuse_PR-134/package/fuse_constraints/test/test_variable_constraints.cpp:117:3:   required from here

[2020-01-22T16:33:28.347Z] /usr/src/googletest/googletest/include/gtest/gtest.h:1392:11: error: comparison between signed and unsigned integer expressions [-Werror=sign-compare]

[2020-01-22T16:33:28.347Z]    if (lhs == rhs) {
```
```
EXPECT_EQ(expected0.size(), std::distance(actual0.begin(), actual0_iter));
```
`std::distance` returns a signed value...which is slightly unexpected to me.

The xenial tests passed, but not the bionic tests. I'm assuming the later compiler version is more strict.

svwilliams commented on fuse_constraints/test/test_variable_constraints.cpp:
```diff
@@ -101,6 +101,23 @@ TEST(VariableConstraints, GetConstraints)
+  auto actual3_iter = actual3.begin();
+  actual3_iter = vars.getConstraints(3u, actual3_iter);
+
+  EXPECT_EQ(expected0.size(), std::distance(actual0.begin(), actual0_iter));
```
`std::distance` returns a signed value. Need a `static_cast` or something to avoid signed-unsigned comparison.

efernandez replied:
Trying with `static_cast<std::ptrdiff_t>`, which I believe is the return type of `std::distance`.

I refrained from using `decltype` because the expression would be large, since `std::distance` is templated and I think the return type depends on the template argument.

efernandez committed:
Cast size() match std::distance type

We're assuming std::distance returns std::ptrdiff_t.
If we don't do this we get a `-Werror=sign-compare` error.
